File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.js
/**
 * @param {string} s
 * @return {number}
 */
var lengthOfLongestSubstring = function(s) {
    // slinding window 
    let left=0;

    // store max length
    let maxlen=0;
    // edge case
    if(s.length<0) return -1;

    // using map
     const map=new Map();

    //  loop
    for(let right=0;right<s.length;right++){
        let char=s[right];

        // check ripeating character
        if(map.has(char) && map.get(char)>=left){
        // window shrink
        left=map.get(char)+1;
        }

       //    update map
         map.set(char,right);

         // find maxlen
       maxlen=Math.max(maxlen,right-left+1);
       }
    
    return maxlen;
};

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        new_set=set()
        left=0
        max_length=0

        for right in range(len(s)):
            while s[right] in new_set:
                new_set.remove(s[left])
                left+=1
            new_set.add(s[right])
            if len(new_set)>max_length:
                max_length=len(new_set)   
        return max_length     